src/mavsdk/plugins/component_metadata_server/component_metadata_server_impl.cpp

```diff
@@ -0,0 +1,163 @@
+#include "component_metadata_server_impl.h"
+#include "mavlink_address.h"
+#include "mavlink_request_message_handler.h"
+#include "callback_list.tpp"
+#include "fs_utils.h"
+#include "crc32.h"
+
+#include <string>
+#include <json/json.h>
+
+namespace mavsdk {
+
+ComponentMetadataServerImpl::ComponentMetadataServerImpl(
+    std::shared_ptr<ServerComponent> server_component) :
+    ServerPluginImplBase(server_component)
+{
+    _server_component_impl->register_plugin(this);
+}
+
+ComponentMetadataServerImpl::~ComponentMetadataServerImpl()
+{
+    _server_component_impl->unregister_plugin(this);
+    if (_metadata_set) {
+        std::filesystem::remove_all(_tmp_path);
+    }
+}
+
+void ComponentMetadataServerImpl::init()
+{
+    _server_component_impl->mavlink_request_message_handler().register_handler(
+        MAVLINK_MSG_ID_COMPONENT_METADATA,
+        [this](
+            uint8_t source_sys_id, uint8_t source_comp_id, MavlinkRequestMessageHandler::Params) {
+            return process_component_metadata_requested();
+        },
+        this);
+}
+
+void ComponentMetadataServerImpl::deinit()
+{
+    _server_component_impl->mavlink_request_message_handler().unregister_all_handlers(this);
+}
+
+std::optional<MAV_RESULT> ComponentMetadataServerImpl::process_component_metadata_requested()
+{
+    LogDebug() << "MAVLINK_MSG_ID_COMPONENT_METADATA request received";
+
+    const std::lock_guard lg{_mutex};
+
+    if (!_metadata_set) {
+        return MAV_RESULT_TEMPORARILY_REJECTED;
+    }
+
+    char general_metadata_uri[MAVLINK_MSG_COMPONENT_METADATA_FIELD_URI_LEN];
+    snprintf(
+        general_metadata_uri, sizeof(general_metadata_uri), "mftp://%s", kComponentGeneralFilename);
+    _server_component_impl->queue_message([&](MavlinkAddress mavlink_address, uint8_t channel) {
+        mavlink_message_t message;
+        mavlink_msg_component_metadata_pack_chan(
+            mavlink_address.system_id,
+            mavlink_address.component_id,
+            channel,
+            &message,
+            _server_component_impl->get_time().elapsed_ms(),
+            _comp_info_general_crc,
+            general_metadata_uri);
+        return message;
+    });
+
+    return MAV_RESULT_ACCEPTED;
+}
+void ComponentMetadataServerImpl::set_metadata(
+    const std::vector<ComponentMetadataServer::Metadata>& metadata)
+{
+    const std::lock_guard lg{_mutex};
+    if (_metadata_set) {
+        LogErr() << "metadata already set";
+        return;
+    }
+
+    // Create tmp directory as ftp root directory
+    const auto tmp_option = create_tmp_directory("mavsdk-component-metadata-server");
+    if (!tmp_option) {
+        LogErr() << "Failed to create tmp directory";
+        return;
+    }
+    _tmp_path = *tmp_option;
+
+    LogDebug() << "Storing metadata under " << _tmp_path;
```

Comment:
By the way, I would probably add a `_debugging` bool like other plugins have it which can be enabled using `export MAVSDK_COMPONENT_METADATA_DEBUGGING=1`, and then you can make it quite verbose in that case.

Reply:
Done in https://github.com/mavlink/MAVSDK/pull/2272/commits/a725299afaeb5f2d30815494174226d285d7f814

Reply:
Very nice!